src/Control/Effect/Interpret.hs

```diff
@@ -1,20 +1,74 @@
-{-# LANGUAGE GeneralizedNewtypeDeriving, MultiParamTypeClasses, RankNTypes, TypeOperators, UndecidableInstances #-}
+{-# language FlexibleContexts #-}
+{-# language FlexibleInstances #-}
+{-# language FunctionalDependencies #-}
+{-# language GeneralizedNewtypeDeriving #-}
+{-# language KindSignatures #-}
+{-# language PolyKinds #-}
+{-# language RankNTypes #-}
+{-# language ScopedTypeVariables #-}
+{-# language TypeApplications #-}
+{-# language TypeOperators #-}
+{-# language UndecidableInstances #-}
+
 module Control.Effect.Interpret
 ( runInterpret
-, InterpretC(..)
 , runInterpretState
-, InterpretStateC(..)
+, InterpretC(..)
 ) where
 
 import Control.Applicative (Alternative(..))
 import Control.Effect.Carrier
-import Control.Effect.Reader
 import Control.Effect.State
 import Control.Monad (MonadPlus(..))
 import Control.Monad.Fail
 import Control.Monad.Fix
 import Control.Monad.IO.Class
 import Control.Monad.Trans.Class
+import Unsafe.Coerce ( unsafeCoerce )
+
+
+newtype InterpretC ( s :: * ) ( sig :: ( * -> * ) -> * -> * ) m a =
+  InterpretC { unInterpretC :: m a }
+  deriving
+    (Functor, Applicative, Monad, MonadFix, MonadFail, MonadIO, MonadPlus, Alternative)
+
+
+instance MonadTrans ( InterpretC s sig ) where
+  lift =
+    InterpretC
+
+
+newtype Handler sig m =
+  Handler { runHandler :: forall s x. sig ( InterpretC s sig m ) x -> InterpretC s sig m x }
+
+
+newtype Tagged a b =
+  Tagged { unTag :: b }
+
+
+class Reifies ( s :: * ) a | s -> a where
+  reflect :: Tagged s a
+
+
+data Skolem
+
+
+newtype Magic a r =
+  Magic ( Reifies Skolem a => Tagged Skolem r )
+
+
+reify :: forall a r. a -> ( forall s. Reifies s a => Tagged s r ) -> r
+reify a k =
+  unsafeCoerce ( Magic @a k ) a
+
+
+instance ( HFunctor eff, HFunctor sig, Reifies s ( Handler eff m ), Monad m, Carrier sig m ) => Carrier ( eff :+: sig ) ( InterpretC s eff m ) where
+  eff ( L eff ) =
+    runHandler ( unTag ( reflect @s ) ) eff
+  eff ( R other ) =
+    InterpretC ( eff ( handleCoercible other ) )
```

Comment:
I think we’re going to want an `INLINE` pragma on this (might be relevant to the benchmarking stuff, I’m not sure).

Reply:
I haven't seen a need for it yet. Should I add it or leave it until benchmarking shows it necessary? Note the benchmarks in my above comment didn't have it, unless you can't reproduce them.

Reply:
Cool, let’s not then!